Add node for custom formatted current DateTime

radumg opened the pull request:
### Purpose

This PR adds a new node to Core.DateTime, called NowByFormat.
The node takes in a string parameter that specifies a format (according to standard MSDN documentation ) and returns an accordingly formatted current system date & time.

Came about as a response to David's question here : https://twitter.com/dourevit/status/817405784367763456

![nowbyformat](https://cloud.githubusercontent.com/assets/11439624/21737448/ae295da2-d470-11e6-8e3e-544713c05a5a.png)

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

Introduces no additional dependencies on Dynamo APIs, external packages, namespaces, etc and does not change the Dynamo API.

### FYIs

ksobon commented:
@radumg great stuff! 

I have only one suggestion. This limits date time formatting to only "now". I think it would be a better node if it had an input for TIME and FORMAT string, and would output a formatted string. Use case is the dropdown node that is part of the UI++ package that lets you select any date/time, not just now. With your addition it would be a single node that fits ALL cases where date/time formatting is needed and hence make it a more usable node.

andydandy74 commented:
Second that

radumg commented:
That's a great suggestion @ksobon ! Will change the node and re-submit PR when ready.
cc @andydandy74

kronz commented:
Thanks @radumg .  We are going to take a look at the code next week and get you some feedback.  I agree with the @ksobon that making this allow a more generalized usage is important to making it as effective as it can be.

kronz commented:
HI @radumg .  Thanks for your patience, here's a couple things.  We are going to be submitting in a Date/Time picker component soon, so it's very appropriate to have your work appear as a more generalized node that accepts a DateTime object and a string, as indicated by @ksobon . 

For the naming and tooltips of this node, I want to just make sure that it aligns with the rest of the library.  The functionality sounds much more like an "action" then a "creation" approach, so we should have the name follow this structure, which doesn't us "by". "DateTime.Format" is probably the simplest.  

Here are my suggestions for ports:
Tooltips for DateTime input:  "DateTime"
Tooltips for Format: "String"
Default value for Format input:  not sure exactly what this should be, but at least something other than what comes out of the Date/Time now by default.  Maybe just the invariant cultural pattern  "MM/dd/yyyy".
Description: This one is a little tricky, as the number of variations are huge.  We should probably have a brief tip and a link to a full description (like https://msdn.microsoft.com/en-us/library/az4se3k1(v=vs.110).aspx).  I suggest: "Formats a Date/Time object by a text description.  Example:  "dd/MM/yyyy".  A full list of formats is at  https://msdn.microsoft.com/en-us/library/az4se3k1(v=vs.110).aspx)"
Search terms: date, time, format, regional 

Thanks for the contribution, I think this node will make a lot of people happy!
-Zach

radumg commented:
Many thanks Zach for the detailed description & great timing too : validates what I've been working on and also answers questions I was thinking of posting (if links allowed in description for example).

For the default value for `format` input string, one thing i'd note is that it might be a good idea to avoid `MM/dd/yyyy` formats as they're not actually culturally universal (despite being invariant) : Europe defaults to `dd/MM/yyyy` for example. To avoid these slight idiosyncrasies, maybe we could use something like `dd MMMM yyyy` ? (prints `01 January 2016`). Let me know what the team would prefer and will use that as default.

kronz commented:
@radumg Your suggestions for default values sound great, I'm happy to go with your modifications.
For the detailed description, if you look at how the Formula node works, we just have an identification of the website, the description doesn't actually carry a hyperlink.  It isn't ideal, but it is effective.
Thanks!  Any idea when you might submit the new code?

radumg commented:
Hi @kronz , refactored this as discussed, with the following note
- had to add a check for Null or Empty format strings to handle case when empty string was supplied from `String` or `Code` nodes (parameter supplied but empty, hence default value is skipped)

I also added a unit test, please have a look and let me know, it's my first foray into UnitTests, so not sure if it's in the right place/format.

Behaviour should be as per screenshot below :
![datetime](https://cloud.githubusercontent.com/assets/11439624/23107794/a3bce124-f6fb-11e6-9cde-5b10f5238bbb.png)

kronz commented:
thanks @radumg .  We'll take a look next week.  Thanks so much, looks perfect from the screenshots.
cc @mjkkirschner @QilongTang 
https://jira.autodesk.com/browse/DYN-446

Racel commented:
@radumg - Looks good! What determined the default format? Maybe we should make it look like one of the standard MSDN formats? https://msdn.microsoft.com/en-us/library/az4se3k1(v=vs.110).aspx

@mjkkirschner @kronz - thoughts?

radumg commented:
Picked that format with a few things in mind
- avoid cultural tropes (us/eu/etc formatting) and localised formats
- be different to results of `DateTime.Now` node (localised)
- showcase other fields you could use, such as the `Day `

Your suggestion @Racel to use a standard format would work very well too. It would be localised though, so there's a decision/preference to be made there.

Out of the standard formats, I would suggest `F`, which results in `Monday, June 15, 2009 1:45:30 PM`. It's a rather complete format and covers points above (shows extra fields `Day` + `seconds` and is different to Now node).

Up to you guys, happy with either, though in hindsight @Racel's suggestion might be better.

Is this being considered for 1.3 ?

Racel commented:
@radumg - I agree with the F format. If there are no objections, I suggest we do that instead. If this change is made, then we we should probably change the description and port tooltip text to match the new format...

And, if we can get these changes in and there are no issues in testing, I think we are trying to squeeze this into 1.3 👍

radumg commented:
hey @Racel , made those quick changes, default is now "F" formatting as below. Also amended the descriptions.
![updated datetime format](https://cloud.githubusercontent.com/assets/11439624/23189744/5c8c79a8-f88c-11e6-9e0c-8ed85f2f9e66.png)

Racel commented:
Looks good @radumg! Thanks for doing that. We are currently in the middle of a sprint, so we probably will get to reviewing this next week. Great work.

QilongTang commented on src/Libraries/CoreNodes/DateTime.cs:
```diff
@@ -50,6 +50,20 @@ public static System.DateTime Today
+        public static string Format(System.DateTime dateTime, string format="F")
+        {
+            if (System.String.IsNullOrEmpty(format)) format = "F";
+            return dateTime.ToString(format, CultureInfo.InvariantCulture);
```
Looks like using of CultureInfo.InvariantCulture makes the format independent with the operating system locale, do you mind me modifying the tooltip for the Format above?

radumg replied:
don't mind, please go ahead @QilongTang 👍
